dummyfs/dummyfs.c

```diff
@@ -511,470 +635,509 @@ int dummyfs_create(void *ctx, oid_t *dir, const char *name, oid_t *oid, unsigned
+
+	if (ei == NULL) {
+		dummyfs_object_put(fs, d);
+		mutexUnlock(fs->mutex);
+		return -ENOENT;
+	}
 
-	d->atime = time(NULL);
+	if ((sizeof(struct dirent) + ei->len + 1) > size) {
+		dummyfs_object_put(fs, d);
+		mutexUnlock(fs->mutex);
+		return -EINVAL;
+	}
 
-	object_unlock(fs, d);
-	object_put(fs, d);
+	d->dir.hint.offs = diroffs;
+	d->dir.hint.entry = ei;
 
-	return ret;
+	dent->d_ino = ei->oid.id;
+	dent->d_reclen = ei->len;
+	dent->d_namlen = ei->len;
+	dent->d_type = ei->type;
+	strcpy(dent->d_name, ei->name);
+
+	dummyfs_object_put(fs, d);
+	mutexUnlock(fs->mutex);
+
+	return 0;
 }
 
 
 int dummyfs_open(void *ctx, oid_t *oid)
 {
+	TRACE;
 	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *o;
 
-	if ((o = dummyfs_get(fs, oid)) == NULL)
+	mutexLock(fs->mutex);
+	dummyfs_object_t *o = dummyfs_object_get(fs, oid);
+	if (o == NULL) {
+		mutexUnlock(fs->mutex);
 		return -ENOENT;
+	}
 
-	object_lock(fs, o);
 	o->atime = time(NULL);
 
-	object_unlock(fs, o);
-	return EOK;
+	mutexUnlock(fs->mutex);
+	return 0;
 }
 
 
 int dummyfs_close(void *ctx, oid_t *oid)
 {
+	TRACE;
 	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *o;
 
-	if ((o = dummyfs_get(fs, oid)) == NULL)
+	mutexLock(fs->mutex);
+	dummyfs_object_t *o = dummyfs_object_get(fs, oid);
+	if (o == NULL) {
+		mutexUnlock(fs->mutex);
 		return -ENOENT;
+	}
 
-	object_lock(fs, o);
 	o->atime = time(NULL);
 
-	object_unlock(fs, o);
-	object_put(fs, o);
-	object_put(fs, o);
+	dummyfs_object_put(fs, o);
+	dummyfs_object_put(fs, o);
+	mutexUnlock(fs->mutex);
 	return EOK;
 }
 
 
-int dummyfs_truncate(void *ctx, oid_t *oid, size_t size)
+int dummyfs_read(void *ctx, oid_t *oid, offs_t offs, char *buff, size_t len)
 {
+	TRACE;
 	dummyfs_t *fs = (dummyfs_t *)ctx;
-	dummyfs_object_t *o;
-	int ret;
-
-	o = object_get(fs, oid->id);
+	size_t cnt = 0;
 
+	mutexLock(fs->mutex);
+	dummyfs_object_t *o = dummyfs_object_get(fs, oid);
 	if (o == NULL) {
+		mutexUnlock(fs->mutex);
 		return -EINVAL;
 	}
 
-	if (!S_ISREG(o->mode)) {
-		if (S_ISDIR(o->mode)) {
-			ret = -EISDIR;
-		}
-		else {
-			ret = -EACCES;
-		}
-		object_put(fs, o);
-		return ret;
+	o->atime = time(NULL);
+
+	if (S_ISDIR(o->mode)) {
+		dummyfs_object_put(fs, o);
+		mutexUnlock(fs->mutex);
+		return -EISDIR;
 	}
 
-	if (o->size == size) {
-		object_put(fs, o);
-		return EOK;
+	if ((offs < 0) || (!S_ISREG(o->mode) && !S_ISLNK(o->mode) && (o->mode != OBJECT_MODE_MEM))) {
+		dummyfs_object_put(fs, o);
+		mutexUnlock(fs->mutex);
+		return -EINVAL;
 	}
 
-	object_lock(fs, o);
+	if (((offs_t)o->size <= offs) || (len == 0)) {
+		dummyfs_object_put(fs, o);
+		mutexUnlock(fs->mutex);
+		return 0;
+	}
 
-	ret = dummyfs_truncate_internal(fs, o, size);
+	size_t left = (len > (o->size - offs)) ? (o->size - offs) : len;
+	if ((o->data == NULL) && (o->chunks == NULL)) {
+		memset(buff, 0, left);
+		cnt = left;
+	}
+	else if ((o->size < DUMMYFS_CHUNKSZ) || (o->mode == OBJECT_MODE_MEM)) {
+		/* Small file */
+		memcpy(buff, o->data + offs, left);
```

Comment:
Is that possible that o->data is NULL at this point? 
(previous condition checks if both o->chunks and o->data is NULL, maybe it's asserted in some other way)

Reply:
`o->chunks` and `o->data` are in union, that's the best that I've came up with to check this. Perhaps `||` would be better

Reply:
After some thought I think that `||` shows the intention better, changed, thanks